library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+     *  explicitly or not.
+     */
+    private def jsName(sym: MethodSymbol): (String, Boolean) = {
+      val annot = memberAnnotations(sym).find(annotIs(_, JSNameAnnotation))
+      val name = annot.flatMap(annotStringArg).getOrElse(defaultName(sym))
+      (name, annot.isDefined)
+    }
+
+    /** Returns all exported members of a type */
+    private def exportedMembers(tpe: Type): Map[String, List[JSMember]] = {
+      val exports = tpe.baseClasses.flatMap(exportedDecls(tpe, _))
+
+      // Group exports by name
+      val lists = for {
+        (name, elems) <- exports.groupBy(_._1)
+      } yield {
+        (name, elems.map(_._2).toList)
+      }
+
+      lists.toMap
+    }
+
+    /** All exported declarations of a class.
+     *  (both @JSExportAll and @JSExport)
+     */
+    private def exportedDecls(origTpe: Type, sym: Symbol) = {
+      require(sym.isClass)
+
+      val exportAll = sym.annotations.exists(annotIs(_, JSExportAllAnnotation))
+
+      for {
+        mem <- sym.info.decls if mem.isMethod && !mem.isConstructor
+        name <- exportNames(mem.asMethod, exportAll)
+      } yield {
+        (name, jsMemberFor(origTpe, mem.asMethod))
+      }
+    }
+
+    /** Get the JS member for a method in [[origTpe]] */
+    private def jsMemberFor(origTpe: Type, sym: MethodSymbol): JSMember = {
+      sym.info.asSeenFrom(origTpe, sym.owner) match {
+        case MethodType(List(param), returnType)
+          if returnType.typeSymbol == definitions.UnitClass &&
+             sym.name.decodedName.toString.endsWith("_=") =>
+          JSSetter(param.info)
+
+        case NullaryMethodType(returnType) =>
+          JSGetter(returnType)
+
+        case info: MethodType =>
+          @tailrec
+          def flatParams(tpe: Type, acc: List[JSMethodParam]): JSMethod = {
+            tpe match {
+              case MethodType(params, returnTpe) =>
+                val ps = params map { p =>
+                  JSMethodParam(p.info, p.asTerm.isParamWithDefault)
+                }
+                flatParams(returnTpe, ps reverse_::: acc)
+              case tpe =>
+                JSMethod(acc.reverse, tpe)
+            }
+          }
+
+          flatParams(info, Nil)
+
+        case PolyType(_, _) =>
+          c.abort(c.enclosingPosition, "Polymorphic methods are currently " +
+            s"not supported. Offending method: ${sym.fullName}")
+
+        case tpe =>
+          sys.error(s"Unexpected method type: $tpe for $sym. Report this as a bug.")
+      }
+    }
+
+    /** Names a method is exported to */
+    private def exportNames(sym: MethodSymbol, exportAll: Boolean) = {
+      lazy val default = defaultName(sym)
+
+      val explicitNames = for {
+        annot <- memberAnnotations(sym)
+        if annotIs(annot, JSExportAnnotation)
+      } yield {
+        annotStringArg(annot).getOrElse(default)
+      }
+
+      if (exportAll && sym.isPublic) default :: explicitNames
```

Comment:
What happens to a method named `apply` with a custom `@JSExport("blahblah")` in the presence of `@JSExportAll`?